### DISCUSSION TITLE
2 questions,1 idea (  Extend: , trait/protocol, with: sugar )

### DISCUSSION THREAD

**Discussion by rd4com** [modular/modular#969] [Ideas]
I know syntactic sugar is not next in the roadmap queue/order, is the``` __enter__``` calling sugar a good idea ?
will mojo have extensions like swift ?
will be able comply a protocol/trait by having a value named x:Int for example? (not just function signatures)
```python
#----------------------syntactic sugar for calling __enter__ ?------------
#problem, cannot indent just for readability
def main():
    h1()
        text("hello world")
        p()
            text("ok")
        _end()
    _end()

#make sense to use with
def main():
    with h1():
        text("ok")
        with p():
            text("ok")
       
#seem ok ?  syntactic sugar for calling __enter__ ?
def main():
    h1():              
        text("ok")
        p():
            text("ok")
        
#--------------------------------------Extensions----------------------------

extend Int:
    fn is_even(self: Int)->Bool:
        return ((self&1)==0)
        
extend PythonObject:
    fn init(inout self: PythonObject, other:my_struct):
        ...

#--------------------------------------protocols/traits----------------------
trait/protocol contains_x_attribute:
    x:Int

```

**Comment by lattner** [comment]
I expect that we'll look into swift-style extensions in the future.  We are also likely to look into "macro" like functionality to allow user-defined statements, subsuming the `with` and `for` and other sugary things into library features.  This is not on the immediate roadmap though.

**Comment by rd4com** [comment]
excellent, keep up the good work i love it!

**Comment by rarebreed** [comment]
As a side commentary to this, I think mojo should call this a `trait` and not a `protocol` since python already has protocols with `typing.Protocol` as a way to deal with structural subtyping for static type checkers.  Having a mojo "protocol" with slightly different semantics from python's protocol will be more confusing IMHO.  Python protocols are slightly confusing already, since there's the ad-hoc definition of "protocol" as in the Iterator protocol, and the more precisely defined protocol from [PEP 544](https://peps.python.org/pep-0544/)

I'm more partial to a rust like `impl`.  I suppose mojo could leverage a python `Protocol` and if it's a `.mojo` file, have an `impl` declaration that actually provides the concrete implementation for the methods.  But I'd still name it something other than _protocol_ to avoid confusion with python's protocol
